iosApp/iosApp/Pages/StopDetails/TripDetailsView.swift

```diff
@@ -18,29 +18,24 @@ struct TripDetailsView: View {
     let stopId: String
     let stopSequence: Int?
 
-    var global: GlobalResponse?
+    var now: Date
 
     @ObservedObject var errorBannerVM: ErrorBannerViewModel
     @ObservedObject var nearbyVM: NearbyViewModel
     @ObservedObject var mapVM: MapViewModel
-
-    @State var tripPredictionsRepository: ITripPredictionsRepository
-    @State var tripRepository: ITripRepository
-    @State var vehicleRepository: IVehicleRepository
-
-    @State var trip: Trip?
-    @State var tripPredictions: PredictionsStreamDataResponse?
-    @State var tripPredictionsLoaded: Bool = false
-    @State var tripSchedulesResponse: TripSchedulesResponse?
-    @State var vehicleResponse: VehicleStreamDataResponse?
-
-    @State var now = Date.now.toKotlinInstant()
+    @ObservedObject var stopDetailsVM: StopDetailsViewModel
 
     let analytics: TripDetailsAnalytics
     let inspection = Inspection<Self>()
 
-    private var routeType: RouteType? {
-        global?.routes[routeId]?.type
+    private var route: Route? {
+        let trip: Trip? = stopDetailsVM.tripPredictions?.trips[tripId]
+        let vehicle: Vehicle? = stopDetailsVM.vehicle
+        return if let routeId = trip?.routeId ?? vehicle?.routeId {
```

Comment:
Good question, testing for a bit without it, it seems like something strange is happening where the trip predictions don't always exist or have the routeId or something, or maybe the wrong predictions are loaded? Though it might just be data issues with certain routes. I'm seeing on the 43, 48, and 8, it's taking a disproportionally long time to load, then if the route isn't pulled from the vehicle, it will display a grey route line and no vehicle card (the card is hidden when `route` is nil). This route is only used for route colors and route type, so it's not actually a problem (yet) if it's finishing a trip on a different route. We could alternately pass those values directly, but doing that for all 3 is a little annoying. After some more testing I'm pretty sure there's something more severe happening here with bus trip predictions, trying to track it down, but I think not falling back to vehicle still has the potential to have confusing and flickery states when the data is weird.

![Simulator Screenshot - iPhone 15 - 2024-12-09 at 12 28 21](https://github.com/user-attachments/assets/f5977841-081a-47e5-855e-841f9fea3459)

Reply:
Okayyy, the issue I was seeing was because the vehicle and trip data loaded by the VM were getting out of sync with the selected trip filter, because there was nothing linking a specific tripFilter value with the loaded trip data, and all the competing `onChange`s and `onDisappears` of different `TripDetailsView`s were conflicting with each other and clearing things that shouldn't be cleared. I did a major refactor of the StopDetailsVM, now all the trip details related responses are stored together in a struct alongside the tripFilter that was initially used to request them. This means that in the UI we can now just check that the current filter is equal to the filter in the trip data, and if not, display the loading state. This also lets us clear and update the trip data in a more self contained way, now the view handles less.

Then once all of that was resolved, I updated the route resolution to also be in the VM and just use the loaded trip object to get the route id, not sure why it was getting the trip from the predictions response before, and now the static trip is always loaded before the predictions or the vehicle, so it's more reliable and won't lead to the behavior in the screenshot in the thread. I still moved the route colors and type out into a new struct because it simplifies consistently setting the fallback values.

Reply:
Okay nice, thanks for digging into the route cause here!